src/QMCWaveFunctions/Fermion/MultiSlaterDeterminantFast.cpp

```diff
@@ -660,117 +674,158 @@ void MultiSlaterDeterminantFast::evaluateDerivatives(ParticleSet& P,
+          for (size_t id = 0; id < Dets.size(); id++)
+          {
+            dhpsi -= v[id];
+          }
           dhpsioverpsi[kk] = dhpsi;
         }
       }
       else
       //usingDETS
       {
-        if (laplSum_up.size() == 0)
-          laplSum_up.resize(Dets[0]->detValues.size());
-        if (laplSum_dn.size() == 0)
-          laplSum_dn.resize(Dets[1]->detValues.size());
-        // assume that evaluateLog has been called in opt routine before
-        //   Dets[0]->evaluateForWalkerMove(P);
-        //   Dets[1]->evaluateForWalkerMove(P);
-        ValueVector_t& detValues_up = Dets[0]->detValues;
-        ValueVector_t& detValues_dn = Dets[1]->detValues;
-        GradMatrix_t& grads_up      = Dets[0]->grads;
-        GradMatrix_t& grads_dn      = Dets[1]->grads;
-        ValueMatrix_t& lapls_up     = Dets[0]->lapls;
-        ValueMatrix_t& lapls_dn     = Dets[1]->lapls;
-        int N1                      = Dets[0]->FirstIndex;
-        int N2                      = Dets[1]->FirstIndex;
-        int NP1                     = Dets[0]->NumPtcls;
-        int NP2                     = Dets[1]->NumPtcls;
-        int n                       = P.getTotalNum();
-        ValueType psiinv            = static_cast<ValueType>(PsiValueType(1.0) / psiCurrent);
-        ValueType lapl_sum          = 0.0;
-        ValueType gg = 0.0, ggP = 0.0;
-        myG_temp = 0.0;
-        ValueVector_t::iterator it(laplSum_up.begin());
-        ValueVector_t::iterator last(laplSum_up.end());
-        ValueType* ptr0 = lapls_up[0];
-        int nels_up     = P.last(0) - P.first(0);
-        while (it != last)
-        {
-          (*it) = 0.0;
-          for (int k = 0; k < nels_up; k++, ptr0++)
-            (*it) += *ptr0;
-          it++;
-        }
-        it          = laplSum_dn.begin();
-        last        = laplSum_dn.end();
-        ptr0        = lapls_dn[0];
-        int nels_dn = P.last(1) - P.first(1);
-        while (it != last)
+        ValueType psiinv   = static_cast<ValueType>(PsiValueType(1.0) / psiCurrent);
+        ValueType lapl_sum = 0.0;
+        const size_t n     = P.getTotalNum();
+        for (size_t id = 0; id < Dets.size(); id++)
         {
-          (*it) = 0.0;
-          for (size_t k = 0; k < nels_dn; k++, ptr0++)
-            (*it) += *ptr0;
-          it++;
+          if (laplSum[id].size() == 0)
+            laplSum[id].resize(Dets[id]->detValues.size());
+          // assume that evaluateLog has been called in opt routine before
+          //   Dets[id]->evaluateForWalkerMove(P);
+          ValueVector_t& detValues_spin = Dets[id]->detValues;
+          ValueMatrix_t& lapls_spin     = Dets[id]->lapls;
+          const size_t n                = P.getTotalNum();
+          ValueVector_t::iterator it(laplSum[id].begin());
+          ValueVector_t::iterator last(laplSum[id].end());
+          ValueType* ptr0 = lapls_spin[0];
```

Comment:
suspicious 0. could you also replace the iterators with direct access?

Reply:
Done